Enhancements to support queue with AIMD and new metrics

spericas opened the pull request:
### Description

Enhancements to support queue with AIMD and new metrics. Issue #9542 and #9304.

### Documentation

In a follow-up PR

VishalSharma-123 commented:
Hi @spericas. 
While conducting the tests, I found out that the data for `number of concurrent requests` and for `RTT Time` was not being recorded properly. It remained at a constant value of `0.0` even under a load of very high user. Upon debugging the code, the `invoke()` method which contains the code for recording the data, was never called. Hence the data was not being recorded. 

For it's fix, I tried to create a private class called `FixedToken` in `FixedLimit.java`, having a similar implementation as the `AIMDToken` in `AIMDLimitImpl.java`, and passing this class in the `QueueSemaphore` constructor for creating the `handler`. The required data for `number of concurrent requests` and for `RTT Time`  is being recorded in `FixedToken`.

With this fix, the metric data for `number of concurrent requests` and for `RTT Time` is being recorded properly now. 

**Output Before the Fix**
`default_fixed_concurrent_requests{mp_scope="vendor",} 0.0
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.5",} 0.0
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.75",} 0.0
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.95",} 0.0
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.98",} 0.0
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.99",} 0.0
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.999",} 0.0
default_fixed_queue_wait_time_seconds_count{mp_scope="vendor",} 0.0
default_fixed_queue_wait_time_seconds_sum{mp_scope="vendor",} 0.0
`

**Output After the Fix**
`default_fixed_concurrent_requests{mp_scope="vendor",} 51.0
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.5",} 0.452722688
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.75",} 0.704380928
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.95",} 1.140588544
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.98",} 1.409024
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.99",} 1.543241728
default_fixed_queue_wait_time_seconds{mp_scope="vendor",quantile="0.999",} 2.281439232
default_fixed_queue_wait_time_seconds_count{mp_scope="vendor",} 10547.0
default_fixed_queue_wait_time_seconds_sum{mp_scope="vendor",}  5652.533172794
`

**Following is the Fix for the Code**

```
private FixedLimit(FixedLimitConfig config) {
        this.config = config;
        this.concurrentRequests = new AtomicInteger();
        this.rejectedRequests = new AtomicInteger();
        this.clock = config.clock().orElseGet(() -> System::nanoTime);
        this.rttTimer = null;
        this.queueWaitTimer = null;

        if (config.permits() == 0 && config.semaphore().isEmpty()) {
            this.semaphore = null;
            this.initialPermits = 0;
            this.queueLength = 0;
            this.handler = new LimitHandlers.NoOpSemaphoreHandler();
        } else {
            this.semaphore = config.semaphore().orElseGet(() -> new Semaphore(config.permits(), config.fair()));
            this.initialPermits = semaphore.availablePermits();
            this.queueLength = Math.max(0, config.queueLength());
            this.handler = new LimitHandlers.QueuedSemaphoreHandler(semaphore,
                                                                    queueLength,
                                                                    config.queueTimeout(),
                                                                    () -> new FixedLimit.FixedToken(clock, concurrentRequests));
        }
    }
```

```
private class FixedToken implements Limit.Token {
        private final long startTime;

        private FixedToken(Supplier<Long> clock, AtomicInteger concurrentRequests) {
            startTime = clock.get();
            concurrentRequests.incrementAndGet();
        }

        @Override
        public void dropped() {
            try {
                updateMetrics(startTime, clock.get());
            } finally {
                FixedLimit.this.semaphore.release();
            }
        }

        @Override
        public void ignore() {
            concurrentRequests.decrementAndGet();
            FixedLimit.this.semaphore.release();
        }

        @Override
        public void success() {
            try {
                updateMetrics(startTime, clock.get());
                concurrentRequests.decrementAndGet();
            } finally {
                FixedLimit.this.semaphore.release();
            }
        }
    }

    void updateMetrics(long startTime, long endTime) {
        long rtt = endTime - startTime;

        if (rttTimer != null) {
            rttTimer.record(rtt, TimeUnit.NANOSECONDS);
        }
    }
```
I have also attached a zip that contains the new `FixedLimit.java` and a `Difference.txt` to highlight the changes.

[Changes for Fixed Limit.zip](https://github.com/user-attachments/files/18486435/Changes.for.Fixed.Limit.zip)

spericas commented:
**Testing report (by L. Govindaraju)**

**Introduction**

In the Initial assessment of Helidon Concurrency Limit Module by running our sample, we noticed small percentage of request breaching KPI leads to higher number of request rejection with 503 due to no queueing support in AIMD and there are no metric exposed for these Concurrency Limit  Components – Fixed and AIMD.
 
Current enhancements in this PR added queueing support for AIMD and below metrics for  Helidon concurrency Limits
- Fixed
  - fixed_queue_length → Number of items waiting in the queue currently
  - fixed_rejected_requests → Number of requests rejected due to Queue Full / Queue timeout
  - fixed_rtt_seconds → Histogram of round trip times(response time) of requests processed
  - fixed_queue_wait_time_seconds →  Histogram of Waiting time of requests in the queue
  - fixed_concurrent_requests → Number of concurrent requests currently processing.
- AIMD
  - aimd_queue_length → Number of items waiting in the queue
  - aimd_rejected_requests → Number of requests rejected due to Queue Full / Queue timeout
  - aimd_rtt_seconds → Histogram of round trip times(response time) of requests processed
  - aimd_queue_wait_time_seconds → Histogram of Waiting time of requests in the queue
  - aimd_concurrent_requests → Number of concurrent requests currently processing
  - aimd_limit → Limit / Permit at the time of querying /metrics
 
The following configurations has been used in this exercise:

- Fixed
      queue-length: 100 
      queue-timeout: "PT1S"
      enable-metrics: true

- AIMD
      timeout: "PT1S"
      backoff-ratio: 0.9
      queue-length: 100
      queue-timeout: "PT1S"
      enable-metrics: true

- Database
          initialPoolSize: 1
          minPoolSize: 5
          maxPoolSize: 10
          connectionWaitTimeout: 10
 
**Metrics Validation**

- Fixed Limit: The values reported by the metrics were verified during multiple round of load tests for both Fixed and AIMD. The reported metrics values are as expected. Enabling the Helidon concurrency Limits metrics, does not cause impact on the performance and resource usage.
 
- AIMD with queuing support: a small percentage of request breaching the KPI, we do not see such huge failures.
Ex: without the queuing, 12% of request breaching the KPI leads to 95% requests rejected with 503, whereas with queuing no requests were rejected.
 
Overall Fixed and AIMD performed similar with now queuing added for AIMD:
1. Example: with permit of 10 (min:5, max:10 for AIMD), we do see avg server time ~107ms. When there are more users than the max permit but there is no significant breach in KPI also we see AIMD and fixed performing similar.
2. Example: For 100 users with 30 permits (min:15, max:30 for AIMD), the avg server time ~318ms for both Fixed as well AIMD. For AIMD, 0.33% of request breached KPI of 1sec. Avg Limit for AIMD was 30.
 
In Burst scenario involving KPI breach, AIMD lower the concurrency limits and queuing is higher for AIMD compared to Fixed during the burst:
1. This leads to better server time for AIMD as fewer concurrent requests contend for the resources.
2. Response time for client is similar when queuing time considered .
3. Ex: with constant 70 users with multiple bursts of addition 30 concurrent users, with 70 permits (min:35, max:70 for AIMD),
```
Avg server time : Fixed – 743ms , AIMD – 528ms
Queuing: Fixed – 1%,  AIMD – 13%
Response Time: Fixed – 744ms, AIMD -738ms
For AIMD, 12% of request breached KPI of 1sec thus lowering the Avg limit to 55.
```
 
In the higher load scenario when there is significant number of requests breaching KPI for AIMD, having the right sizing for queue parameters is critical.
1. Ex: with queue size of 100 and queue timeout of 1sec,
For 200 users with 100 permits (min:50, max:100 for AIMD),
Fixed had 10.5% request rejected due to queue timeout of 1sec with avg server time of 1062ms.
AIMD had 97% request rejected due to queue being full with avg server time of 744ms.
For AIMD, avg Limit was 55 as 21% of request breached KPI of 1sec.
2. Ex: with queue size of 100 and queue timeout of 2 sec,
For 200 users with 100 permits (min:50, max:100 for AIMD),
Fixed had no errors with avg server time of 1062ms.
AIMD had 98% request rejected due to queue being full with avg server time of 745ms
For AIMD, avg Limit was 57 as 22% of request breached KPI of 2 sec.
3. Ex: with queue size of 200 and queue timeout of 1 sec,
For 200 users with 100 permits (min:50, max:100 for AIMD),
Fixed had 8.9% request rejected due to queue timeout of 1sec with avg server time of 1062ms.
In AIMD requested rejected reduced to 33% due to queue timeout of 1sec with avg server time of 699ms.
For AIMD, avg Limit was 65 as 17.9% of request breached KPI of 1sec.
4. Ex: with queue size of 200 and queue timeout of 2 sec,
For 200 users with 100 permits (min:50, max:100 for AIMD),
Fixed had no errors with avg server time of 1063ms.
AIMD had no errors with avg server time of 708ms.
For AIMD, avg Limit was 65 as 18.3% of request breached KPI of 2 sec.
 
In the negative scenario of poorly performing application where every request breaches the KPI of 1 sec due to bad code.
Ex: Application have poorly written SQL which breaches the KPI of 1 sec in every request.
For 30 users with 30 permits (min:15, max:30 for AIMD), queue timeout of 1 sec:

- Fixed had 0.2% request failed with 500 due to request unable to get connection from DB pool after waiting for max wait time of 10 seconds.
- AIMD 46% of request rejected due to queue timeout of 1 sec.

Avg server time: Fixed- 3313ms, AIMD- 1660ms.
For AIMD, avg Limit was 15(min)  as 100% of request breached KPI of 1sec.

For 30 users with 30 permits (min:15, max:30 for AIMD), queue timeout of 2sec,

- Fixed had 0.2% request failed with 500 due to request unable to get connection from DB pool after waiting for max wait time of 10 seconds.
- AIMD, request rejected reduced to 9% of due to higher queue timeout of 2 sec.

Avg server time: Fixed- 3310ms, AIMD- 1661ms.
For AIMD, avg Limit was 15(min)  as 100% of request breached KPI of 1sec.

tjquinno reviewed (approved):
I focused on the metrics part. LGTM.

spericas commented:
In addition to Tim's approval, this new feature has been tested by PSR and a report above shows its outcome. See also #9725.
